lib/dataset/bin.cpp

```diff
@@ -299,9 +299,18 @@ class TargetBinBuilder {
         } else {
           update_indices_by_binning(indices, get_coord(dim), key, linspace);
         }
-      } else if (action == AxisAction::Existing)
-        update_indices_from_existing(indices, dim);
-      else if (action == AxisAction::Join) {
+      } else if (action == AxisAction::Existing) {
+        // Similar to binning along an existing dim, if a dimension is simply
+        // kept unchanged there is a 1:1 mapping from input to output dims. We
+        // can thus avoid storing and processing a lot of length-0 contributions
+        // to bins.
+        if (m_offsets.dims().empty() && m_dims[dim] != 0) {
+          m_nbin = (m_dims.volume() / m_dims[dim]) * units::one;
+          m_offsets = make_range(0, m_dims[dim], 1, dim);
+        } else {
+          update_indices_from_existing(indices, dim);
+        }
```

Comment:
Are both cases covered by tests?

I am confused why one branch sets members of the `TargetBinBuilder` and the other only modifies the indices.

Reply:
Not sufficiently, and indeed there appears to be an issue with this in a case I had not considered.

Reply:
@jl-wynen Can you have another look? Cheers!